tests/integration_tests/functional/test_api.py

```diff
@@ -1342,6 +1342,142 @@ def test_api_balloon(test_microvm_with_api):
+        'vcpu_count': 2,
+        'mem_size_mib': 256,
+        'smt': True,
+        'track_dirty_pages': False,
+        'cpu_template': 'C3',
+    }
+    setup_cfg['boot-source'] = {
+        'kernel_image_path': '/vmlinux.bin',
+        'initrd_path': None
+    }
+    setup_cfg['drives'] = [{
+        'drive_id': 'rootfs',
+        'path_on_host': '/bionic.rootfs.ext4',
+        'is_root_device': True,
+        'partuuid': None,
+        'is_read_only': False,
+        'cache_type': 'Unsafe',
+        'rate_limiter': None,
+        'io_engine': 'Sync'
+    }]
+
+    # Add a memory balloon device.
+    response = test_microvm.balloon.put(amount_mib=1, deflate_on_oom=True)
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    setup_cfg['balloon'] = {
+        'amount_mib': 1,
+        'deflate_on_oom': True,
+        'stats_polling_interval_s': 0
+    }
+
+    # Add a vsock device.
+    response = test_microvm.vsock.put(
+        guest_cid=15,
+        uds_path='vsock.sock'
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    setup_cfg['vsock'] = {
+        'guest_cid': 15,
+        'uds_path': 'vsock.sock'
+    }
+
+    # Add a net device.
+    iface_id = '1'
+    tapname = test_microvm.id[:8] + 'tap' + iface_id
+    tap1 = net_tools.Tap(tapname, test_microvm.jailer.netns)
+    guest_mac = '06:00:00:00:00:01'
+    tx_rl = {
+        'bandwidth': {
+            'size': 1000000,
+            'refill_time': 100,
+            'one_time_burst': None
+        },
+        'ops': None
+    }
+    response = test_microvm.network.put(
+        iface_id=iface_id,
+        guest_mac=guest_mac,
+        host_dev_name=tap1.name,
+        tx_rate_limiter=tx_rl
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    setup_cfg['network-interfaces'] = [{
+        'iface_id': iface_id,
+        'host_dev_name': tap1.name,
+        'guest_mac': '06:00:00:00:00:01',
+        'rx_rate_limiter': None,
+        'tx_rate_limiter': tx_rl
+    }]
+
+    # Update MMDS config.
+    mmds_config = {
+        'version': 'V2',
+        'ipv4_address': '169.254.169.250',
+        'network_interfaces': ['1']
+    }
+    response = test_microvm.mmds.put_config(json=mmds_config)
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    setup_cfg['logger'] = None
+    setup_cfg['metrics'] = None
+    setup_cfg['mmds-config'] = {
+        'version': 'V2',
+        'ipv4_address': '169.254.169.250',
+        'network_interfaces': ['1']
+    }
+
+    # Start the microvm.
+    test_microvm.start()
+
+    microvm_builder = MicrovmBuilder(bin_cloner_path)
+
+    root_disk = test_microvm.vm.disks[0]
+    ssh_key = test_microvm.ssh_key
+
+    # Create a snapshot builder from a microvm.
+    snapshot_builder = SnapshotBuilder(test_microvm)
+
+    # Create base snapshot.
+    snapshot = snapshot_builder.create([root_disk.local_path()],
```

Comment:
@alindima I'm a bit confused on this final test. I've been trying to find tests to reference for creating a microvm from a snapshot (e.g. the tests in `test_snapshot_basic.py`) and they mostly seem to use context to get the `ssh_key`, `root_disk`, and `snapshot_type` but when I try to pass `context` into the new test I made, I get an error that I can't get past. This approach that I'm trying is using the `MicrovmBuilder` and `SnapshotBuilder` but I also tried another approach where I was trying to use `SnapshotHelper.load` but that failed as well. Any tips on which approach would be the best? Thanks!

Reply:
This is a little tricky to get right. Try applying this patch:
```
diff --git a/tests/integration_tests/functional/test_api.py b/tests/integration_tests/functional/test_api.py
index be039c84..a0b66a84 100644
--- a/tests/integration_tests/functional/test_api.py
+++ b/tests/integration_tests/functional/test_api.py
@@ -19,7 +19,7 @@ import host_tools.network as net_tools
 from conftest import _test_images_s3_bucket
 
 from framework.utils import is_io_uring_supported
-from framework.artifacts import ArtifactCollection, SnapshotType
+from framework.artifacts import ArtifactCollection, SnapshotType, NetIfaceConfig, DEFAULT_DEV_NAME, DEFAULT_TAP_NAME
 from framework.builder import MicrovmBuilder, SnapshotBuilder
 
 MEM_LIMIT = 1000000000
@@ -1342,19 +1342,24 @@ def test_api_balloon(test_microvm_with_api):
     assert test_microvm.api_session.is_status_bad_request(response.status_code)
 
 
-def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_cloner_path):
+def test_get_full_config_after_restoring_snapshot(bin_cloner_path):
     """
     Test the reported configuration of a microVM configured with all resources.
 
     @type: functional
     """
-    test_microvm = test_microvm_with_api
+    microvm_builder = MicrovmBuilder(bin_cloner_path)
+    test_microvm = microvm_builder
+
+    net_iface = NetIfaceConfig()
+    vm_instance = microvm_builder.build_vm_nano(net_ifaces=[net_iface])
+    test_microvm = vm_instance.vm
+    root_disk = vm_instance.disks[0]
+    ssh_key = vm_instance.ssh_key
 
     setup_cfg = {}
 
-    test_microvm.spawn()
     # Basic config also implies a root block device.
-    test_microvm.basic_config(rootfs_io_engine="Sync")
     setup_cfg['machine-config'] = {
         'vcpu_count': 2,
         'mem_size_mib': 256,
@@ -1362,13 +1367,10 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         'track_dirty_pages': False,
         'cpu_template': 'C3',
     }
-    setup_cfg['boot-source'] = {
-        'kernel_image_path': '/vmlinux.bin',
-        'initrd_path': None
-    }
+    test_microvm.machine_cfg.patch(**setup_cfg['machine-config'])
     setup_cfg['drives'] = [{
         'drive_id': 'rootfs',
-        'path_on_host': '/bionic.rootfs.ext4',
+        'path_on_host': f"/{os.path.basename(root_disk.local_path())}",
         'is_root_device': True,
         'partuuid': None,
         'is_read_only': False,
@@ -1397,11 +1399,20 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         'uds_path': 'vsock.sock'
     }
 
-    # Add a net device.
-    iface_id = '1'
-    tapname = test_microvm.id[:8] + 'tap' + iface_id
-    tap1 = net_tools.Tap(tapname, test_microvm.jailer.netns)
-    guest_mac = '06:00:00:00:00:01'
+    setup_cfg['logger'] = None
+    setup_cfg['metrics'] = None
+    setup_cfg['mmds-config'] = {
+        'version': "V1",
+        'network_interfaces': [DEFAULT_DEV_NAME]
+    }
+
+    response = test_microvm.mmds.put_config(json=setup_cfg['mmds-config'])
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    # Start the microvm.
+    test_microvm.start()
+
+    # Add a tx rate limiter to the net device.
     tx_rl = {
         'bandwidth': {
             'size': 1000000,
@@ -1410,46 +1421,19 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
         },
         'ops': None
     }
-    response = test_microvm.network.put(
-        iface_id=iface_id,
-        guest_mac=guest_mac,
-        host_dev_name=tap1.name,
+    response = test_microvm.network.patch(
+        iface_id=DEFAULT_DEV_NAME,
         tx_rate_limiter=tx_rl
     )
     assert test_microvm.api_session.is_status_no_content(response.status_code)
     setup_cfg['network-interfaces'] = [{
-        'iface_id': iface_id,
-        'host_dev_name': tap1.name,
-        'guest_mac': '06:00:00:00:00:01',
+        'guest_mac': net_tools.mac_from_ip(net_iface.guest_ip),
+        'iface_id': DEFAULT_DEV_NAME,
+        'host_dev_name': DEFAULT_TAP_NAME,
         'rx_rate_limiter': None,
         'tx_rate_limiter': tx_rl
     }]
 
-    # Update MMDS config.
-    mmds_config = {
-        'version': 'V2',
-        'ipv4_address': '169.254.169.250',
-        'network_interfaces': ['1']
-    }
-    response = test_microvm.mmds.put_config(json=mmds_config)
-    assert test_microvm.api_session.is_status_no_content(response.status_code)
-
-    setup_cfg['logger'] = None
-    setup_cfg['metrics'] = None
-    setup_cfg['mmds-config'] = {
-        'version': 'V2',
-        'ipv4_address': '169.254.169.250',
-        'network_interfaces': ['1']
-    }
-
-    # Start the microvm.
-    test_microvm.start()
-
-    microvm_builder = MicrovmBuilder(bin_cloner_path)
-
-    root_disk = test_microvm.vm.disks[0]
-    ssh_key = test_microvm.ssh_key
-
     # Create a snapshot builder from a microvm.
     snapshot_builder = SnapshotBuilder(test_microvm)
 
@@ -1465,11 +1449,12 @@ def test_get_full_config_after_restoring_snapshot(test_microvm_with_api, bin_clo
     # We expect boot-source, machine-config.smt, and machine-config.cpu_template to all
     # be empty/default after restoring from a snapshot.
     expected_cfg['boot-source'] = {
-        'kernel_image_path': None,
+        'kernel_image_path': '',
         'initrd_path': None
     }
     expected_cfg['machine-config']['smt'] = False
-    expected_cfg['machine-config']['cpu_template'] = None
+    expected_cfg['machine-config'].pop('cpu_template')
+    expected_cfg['mmds-config'] = None
 
     # Validate full vm configuration post-restore.
     response = microvm.full_cfg.get()
```

Reply:
Notice that the mmds_config is still reported as empty after snapshotting, you need to also hook that in when re-adding the network device to VmResources.
There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

Reply:
Unfortunately, some stuff came up in my personal life. I'll look back into this soon (should be within a few days).

Reply:
Sure, no worries!

Reply:
Sorry I'm struggling so much on this final test. I'm not sure what you mean with the following:

> you need to also hook that in when re-adding the network device to VmResources.

> There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

It seems that the MMDS config always shows as null (which is maybe what you're saying)? If it should be null after snapshot, then I'm thinking what I have currently, is right.

Also, any idea what I can use to test the `cpu_template` of the `machine-config`? I had it set to `C3` because that's what some other tests were using but I was getting an `InvalidVendor` that was caused by the C3 `cpu_template` so I ended up getting rid of setting the `cpu_template` (at least temporarily). 

As a side not, I'll merge the 2 final tests commits together this is worked through. I just wanted to push what I had for review.

Reply:
> Sorry I'm struggling so much on this final test.

No worries. Don't get discouraged, you've done the more difficult part 😄 .

> It seems that the MMDS config always shows as null (which is maybe what you're saying)?

Yes, that what I was referring to. It should not be empty after restoring from snapshot. The `NetState` struct contains the optional `MmdsNetworkStackState` which can be used to extract the Mmds config. This should be done in the `update_from_restored_device` function, for each of the net devices.

> There is a known limitation that MMDS version 2 is lost after snapshot, so use the default value in that case.

What I meant by this is that the used MMDS version is not yet saved in the snapshot (so it should always be V1, the default).

Reply:
> Also, any idea what I can use to test the cpu_template of the machine-config? I had it set to C3 because that's what some other tests were using but I was getting an InvalidVendor that was caused by the C3 cpu_template so I ended up getting rid of setting the cpu_template (at least temporarily).

Hmm, are you using an AMD host by any chance?

Reply:
You should use the `utils.get_cpu_vendor()` in the test, and only set that CPU template for intel hosts

Reply:
> Hmm, are you using an AMD host by any chance?

Yes I am actually. I'll use your recommendation of using `utils.get_cpu_vendor` to conditionally set the `cpu_template`. Thanks!

After rereading a bunch of MMDS documentation, reading through the code for a while, and experimenting with the API, I feel like I'm understanding the MMDS change more than I was before. A few implementation questions I still have though are:

1. You mentioned using `MmdsNetworkStackState` in the `NetState`. I noticed that property is currently not public; are we comfortable making it public? On a related note though, the `update_from_restored_device` currently doesn't have access to the `NetState` but it does have access to the `Net` device which seems to have the MMDS network stack stored on it so I'm thinking about using that instead of the `NetState`. Does that sound reasonable or is there a reason I should still make use `NetState`? If it sounds reasonable to use `Net` instead of `NetState`, then ignore my first question about making the `NetState` MMDS state public. 
2. There is only one `ipv4_address` that exists for the `MmdsConfig`, however, each `Net` device has a `ipv4_addr` that seems to refer to the MMDS server. Presumably, the `ipv4_addr` for all net devices will always be the same (since there is only a single MMDS server) but correct me if that presumption is wrong. What I'm thinking about doing is setting the `MmdsConfig` ipv4 address to the default (`169.254.169.254`) in `build_microvm_from_snapshot` and then updating it to the value of the `ipv4_addr` of the net device in `update_from_restored_device`. Does that sound reasonable? (see the code snippet below for an example of how I might go about the update in `update_from_restored_device`)

```rs
 SharedDeviceType::SharedNetwork(network) => {
    if let Some(mmds_ns) = &network.lock().unwrap().mmds_ns {
        let mmds_config = self.mmds_config.as_mut().unwrap();
        mmds_config.ipv4_address = Some(mmds_ns.ipv4_addr);
        mmds_config
            .network_interfaces
            .push(network.lock().expect("Poisoned lock").id().to_owned());
    }

    self.net_builder.add_device(network);
}
```

Reply:
> You mentioned using MmdsNetworkStackState in the NetState. I noticed that property is currently not public; are we comfortable making it public? On a related note though, the update_from_restored_device currently doesn't have access to the NetState but it does have access to the Net device which seems to have the MMDS network stack stored on it so I'm thinking about using that instead of the NetState. Does that sound reasonable or is there a reason I should still make use NetState? If it sounds reasonable to use Net instead of NetState, then ignore my first question about making the NetState MMDS state public.

This was a mistake from me, I was indeed referring to the `Net` struct.

> There is only one ipv4_address that exists for the MmdsConfig, however, each Net device has a ipv4_addr that seems to refer to the MMDS server. Presumably, the ipv4_addr for all net devices will always be the same (since there is only a single MMDS server) but correct me if that presumption is wrong.

This is true 👍🏻 

> What I'm thinking about doing is setting the MmdsConfig ipv4 address to the default (169.254.169.254) in build_microvm_from_snapshot and then updating it to the value of the ipv4_addr of the net device in update_from_restored_device. Does that sound reasonable? (see the code snippet below for an example of how I might go about the update in update_from_restored_device)

No need to set the ipv4 address to default before `update_from_restored_device` (since the `ipv4_addr` field on the `MmdsNetworkStack` is not optional; if there is no network interface with MMDS configured, then there is no MMDS configuration at all). The code snippet you pasted looks good to me overall.

Note that `self.mmds_config.as_mut().unwrap()` may panic as it's written in the snippet.